text/0219-ecs-service-extensions.md

```diff
@@ -0,0 +1,503 @@
+  in the class name customers just create a service, and add it to an environment. The
+  service automatically modifies its settings as needed to adjust to the capacity available
+  in the environment.
+
+# Design Summary
+
+In this proposal we introduce two new constructs:
+
+- `Environment` - A simple construct which wraps up a VPC, cluster, and other top level
+  resources needed for ECS services to run on an AWS account
+- `Service` - A single application that runs inside an `Environment`
+
+And two new associated foundational classes used by the `Service` construct:
+
+- `ServiceDescription` - A class which is utilized to build out a description of the
+  application and the features that should be attached to the application. This description
+  is consumed by the `Service` class
+- `ServiceExtension` - An abstract class for a single type of feature that you want to
+  attach to a service. A `ServiceExtension` is added to a `ServiceDescription` and then
+  the `ServiceDescription` is used to make a new `Service` construct.
+
+The following code example summarizes how you would use these classes together:
+
+```js
+const myServiceDesc = new ServiceDescription();
+myServiceDesc.add(new Container({
+  cpu: 1024,
+  memoryMiB: 2048,
+  trafficPort: 80,
+  image: ContainerImage.fromRegistry('nathanpeck/greeter')
+}));
+myServiceDesc.add(new AppMeshExtension({ mesh }));
+myServiceDesc.add(new FireLensExtension());
+myServiceDesc.add(new XRayExtension());
+myServiceDesc.add(new CloudwatchAgentExtension());
+myServiceDesc.add(new HttpLoadBalancerExtension());
+
+const myEnvironment = new Environment(stack, 'production', {
+  vpc: vpc,
+  cluster: cluster,
+  capacityType: capacityTypes.EC2 || capacityTypes.FARGATE
+});
+
+const myService = new Service(stack, 'my-service', {
+  environment: myEnvironment
+  serviceDescription: myServiceDesc
+});
+```
+
+# Detailed Design
+
+Let's take a more detailed look at each class:
+
+## Class: `ServiceDescription`
+
+This class is not an actual construct, it is just a helper class which is used to build
+out a description of the service to build. This class is used because traditionally all
+CDK constructs are immutable. Once a `Service` is created using its constructor most if
+not all of its properties should be immutable. Therefore its necessary to build out a
+full list of all the things that need to be constructed, prior to construction time.
+
+The `ServiceDescription` class has one method: `ServiceDescription.add(ServiceExtension)`.
+Usage looks similar to this:
+
+```js
+const myServiceDesc = new ServiceDescription();
+myServiceDesc.add(new Container();
+myServiceDesc.add(new AppMeshExtension({ mesh }));
+myServiceDesc.add(new FireLensExtension());
+myServiceDesc.add(new XRayExtension());
+```
+
+## Class: `ServiceExtension`
+
+This class defines an actual extension itself. The extension is collection of hooks which
```

Comment:
Why did you chose an "inversion of control" pattern (hooks) over simple mutation API for Service? IOC is considered harder to reason about and by definition very "leaky" and prone to mistakes. Extensions can easily run over each other without knowledge.

I can see why this might be appealing at first glance but our experience shows that it's worth to actually expose appropriate degrees of freedom (mutation API) in order to allow the object being mutated to maintain it's invariants and perform validations.

Reply:
There are a few reasons:

- Not everything in the underlying L2 constructs is mutable. For example once a `TaskDefinition` is created it is no longer possible to modify the `proxyConfiguration`. That can only be set at the time of construction. This is an example of where immutability of certain L2 properties requires the `Service` class to have preknowledge of all settings prior to creating the underlying L2 resource. We can't just create the underlying L2 resources first and then let extensions change them after it is created, in part because it isn't possible for all properties.
- The hook approach enforces and encourages best practices. For example the `resolveContainerDependencies` hook encourages all extension owners to think about how their extension's container relates to other extension containers and what other containers it might need to wait on during task start phase. With an unstructured approach of mutability there would be no gentle encouragement of an unimplemented hook waiting for the extension owner to fill out with their own logic. It would be too easy to implement an extension that is not fully featured and does not account for container dependencies. This applies to other hooks as well.
- Inversion of control gives extensions a way to reason about what other extensions are added. If extensions are unaware of each other and just changing an underlying construct directly they only see each other's changes after the fact, and aren't fully aware that the other extension exists. This inversion of control enforces that all extensions have been added to the service first, and are able to know of each other prior to the extension's mutation logic running. Without inversion of control an extension can start freely mutating underlying resources prematurely, making decisions that do not account for the future addition of another extension that would otherwise have changed the behavior of the first extension. For that reason direct control would require users to reason about the order that they add extensions, vs inversion of control which allows users to add all their extensions in any order, and at service construction time every extension is able to access the full picture to make independent decisions at the time that each hook is run

To summarize I felt that inversion of control would provide an opinionated, structured approach that makes it easier to implement an extension that behaves correctly, and it would allow us to maintain immutability of the underlying construct resources. I'd love to chat more on this subject and see if I'm missing an easy way to solve these problems without inversion of control.